File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/assignment/others/basedcyclic/BestCyclicAssignment.java
package cujae.inf.ic.om.assignment.others.basedcyclic;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.solution.Cluster;
import cujae.inf.ic.om.problem.solution.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

/*Clase que modela como asignar el mejor cliente al último cliente - depósito asignando en forma paralela */
public class BestCyclicAssignment extends ByNotUrgency {

	private Solution solution = new Solution();
	
	private ArrayList<Cluster> clusters;
	private ArrayList<Customer> customersToAssign;
	private NumericMatrix costMatrix;
    	
	public BestCyclicAssignment() {
		super();
	}

	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}
	
	@Override
	public void initialize() {
    	clusters = initializeClusters();
		customersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		costMatrix = initializeCostMatrix(customersToAssign, Problem.getProblem().getDepots(), distanceType);
	}
	
	@Override
    public void assign() {
		int posElementMatrix = -1;
	    
		double capacityDepot = 0.0;   
	    
	    double requestCustomer = 0.0;    
	    
	    double requestCluster = 0.0; 
	    int posCluster = -1;
	    
	    RowCol rcBestAllSelected = null;
		
		int totalItems = customersToAssign.size();
		int totalClusters = Problem.getProblem().getDepots().size();
		
		ArrayList<Integer> itemsSelected = new ArrayList<>();
		
		for(int i = 0; i < totalClusters; i++) 
        	itemsSelected.add((totalItems + i));
		
        while((!customersToAssign.isEmpty()) && (!clusters.isEmpty()))
        {
        	rcBestAllSelected = getBestValueOfSelected(itemsSelected, costMatrix, totalItems);
        	posElementMatrix = rcBestAllSelected.getRow();
        	
        	if(posElementMatrix >= totalItems) 
        		posCluster = posElementMatrix - totalItems;
        	else
        		posCluster = findElementInSelected(itemsSelected, posElementMatrix);

        	capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(clusters.get(posCluster).getIDCluster()));
        		    	
            requestCustomer = Problem.getProblem().getCustomers().get(rcBestAllSelected.getCol()).getRequestCustomer();
        
			if(posCluster != -1)
			{
				requestCluster = clusters.get(posCluster).getRequestCluster();
				
				if (capacityDepot >= requestCluster + requestCustomer) 
				{	
					requestCluster += requestCustomer;
					
					clusters.get(posCluster).setRequestCluster(requestCluster);
					clusters.get(posCluster).getItemsOfCluster().add(Problem.getProblem().getCustomers().get(rcBestAllSelected.getCol()).getIDCustomer());
					
					customersToAssign.remove(Problem.getProblem().findPosCustomer(customersToAssign, Problem.getProblem().getCustomers().get(rcBestAllSelected.getCol()).getIDCustomer()));
	
					costMatrix.fillValue(0, rcBestAllSelected.getCol(), (totalItems  + totalClusters - 1), rcBestAllSelected.getCol(), Double.POSITIVE_INFINITY);
					costMatrix.fillValue(posElementMatrix, 0, posElementMatrix, (totalItems - 1), Double.POSITIVE_INFINITY);
					
					itemsSelected.remove(posCluster);
					itemsSelected.add(posCluster, rcBestAllSelected.getCol());
				} 
				else 
					costMatrix.setItem(posElementMatrix, rcBestAllSelected.getCol(), Double.POSITIVE_INFINITY);
				
				if(isFullDepot(customersToAssign, requestCluster, capacityDepot))
				{
					if(!(clusters.get(posCluster).getItemsOfCluster().isEmpty()))
						solution.getClusters().add(clusters.get(posCluster));

					clusters.remove(posCluster);
					clusters.add(posCluster, null);
					
					itemsSelected.remove(posCluster);	
					itemsSelected.add(posCluster, -1);
    				costMatrix.fillValue(posElementMatrix, 0, rcBestAllSelected.getRow(), (totalItems - 1), Double.POSITIVE_INFINITY);
				}
			}
        }
	}
	
    @Override
    public Solution finish() {   	
		if(!customersToAssign.isEmpty())					
			for(int i = 0; i < customersToAssign.size(); i++)	
				solution.getUnassignedItems().add(customersToAssign.get(i).getIDCustomer());
				
		if(!clusters.isEmpty()) //no va a pasar
			for(int k = 0; k < clusters.size(); k++)
				if((clusters.get(k) != null) && (!clusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(clusters.get(k));
		
		OSRMService.clearDistanceCache();
		
        return solution;
    }    
	
	private RowCol getBestValueOfSelected(ArrayList<Integer> itemsSelected, NumericMatrix costMatrix, int totalItems){
		RowCol rcBestAll = null;
		
		if((itemsSelected != null) && (!itemsSelected.isEmpty())) //esto no pasa
		{	// verificar que no tome valores de la diagonal principal
			
			RowCol rcCurrent = null;
			boolean isFirst = true;
			
			//rcBestAll = costMatrix.indexLowerValue(itemsSelected.get(0), 0, itemsSelected.get(0), (totalItems - 1)); 
			//función que busque la primera posicion que no es -1 en itemsSelected
			for(int i = 0; i < itemsSelected.size(); i++)
			{
				if(itemsSelected.get(i) != -1)
				{
					if(isFirst)
					{
						rcBestAll = costMatrix.indexLowerValue(itemsSelected.get(i), 0, itemsSelected.get(i), (totalItems - 1));
						rcCurrent = rcBestAll;
						
						isFirst = false;
					}		
					else
						rcCurrent = costMatrix.indexLowerValue(itemsSelected.get(i), 0, itemsSelected.get(i), (totalItems - 1));
					
					if(costMatrix.getItem(rcCurrent.getRow(), rcCurrent.getCol()) < costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()))
						rcBestAll = rcCurrent;
				}			
			}
		}
		return rcBestAll;
	}
	
	private int findElementInSelected(ArrayList<Integer> itemsSelected, Integer idItem){
		int posCluster = -1;
		
		int i = 0;
		boolean found = false;
		
		while((i < itemsSelected.size()) && (!found))
		{
			if(itemsSelected.get(i).intValue() == idItem.intValue())
			{
				posCluster = i;
				found = true;
			}
			else
				i++;
		}
		return posCluster;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/others/basedcyclic/best_cyclic_assignment.py
import numpy as np
from typing import List, Optional, Tuple
from ...classical.by_not_urgency import ByNotUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.output.solution import Solution
from ....problem.output.cluster import Cluster

class BestCyclicAssignment(ByNotUrgency):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        
    def to_clustering(self) -> Solution:
        self.initialize()
        self.assign()
        return self.finish()   
        
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.total_clusters = len(Problem.get_problem().get_depots())
        
    def assign(self):
        cost_matrix = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)

        is_first: bool = True
        last_depot_used_id = -1
        
        while any(self.list_customers_to_assign) and self.list_clusters:
            if is_first:
                min_value = np.min(cost_matrix) 
                row_best, col_best = np.where(cost_matrix == min_value)
                row_best, col_best = row_best[0], col_best[0]
                is_first = False
            else:
                col = cost_matrix[last_depot_used_id, :]
                col_best = np.argmin(col)
                row = cost_matrix[:, col_best]
                row_best = np.argmin(row)

            selected_customer = Problem.get_problem().get_customers()[col_best]
            request_customer = selected_customer.get_request_customer()
              
            id_depot = Problem.get_problem().get_list_id_depots()[row_best]    
            selected_depot: Depot = Problem.get_problem().get_depot_by_id_depot(id_depot)
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(selected_depot)
                
            pos_cluster = self.find_cluster(id_depot, self.list_clusters)
            cluster = self.list_clusters[pos_cluster]
            
            if pos_cluster != -1:
                request_cluster = cluster.get_request_cluster()

                if capacity_depot >= (request_cluster + request_customer):
                    cluster.set_request_cluster(request_cluster + request_customer)
                    cluster.get_items_of_cluster().append(selected_customer.get_id_customer())

                    last_depot_used_id = row_best
                    self.list_customers_to_assign[col_best] = None
                    cost_matrix[:, col_best] = float('inf')
                else:
                    cost_matrix[row_best, col_best] = float('inf')
                    
                if self.is_full_depot(self.list_customers_to_assign, cluster.get_request_cluster(), capacity_depot):
                    if cluster.get_items_of_cluster():
                        self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                    
    def finish(self):
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())

        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
        
        OSRMService.clear_distance_cache()
        
        return self.solution